enhancements/olm/platform-operators.md

```diff
@@ -0,0 +1,532 @@
+
+The PO manager would abstract the process of sourcing registry+v1 bundles from catalog(s), choosing the "best bundle" to install for each desired package, and ensuring those packages get installed on the cluster.
+
+In order to determine the set of bundles that need to be installed, the PO manager can interact with the default, `redhat-operators` CatalogSource, and use OLM's existing registry gRPC APIs to list the available bundles from those catalog contents.
+
+> The OLM marketplace component is now an optional CVO component. Admins can disable this component using the component capabilities functionality introduced in previous 4.x releases. Available workarounds include admins creating their own CatalogSource resources. This behavior may require explicit documentation, or expanding the phase 0 design scope to be able to accommodate all the catalog sources in the cluster.
+
+After determining the set of bundles that need to exist in order to satisfy admin package intent, the PO manager's application layer will need to attempt to install those bundles. The management of those bundle contents will be delegated to OLM's rukpak layer, and the POM component would be responsible for generating an underlying BundleDeployment resource for the bundle that was selected. This underlying BundleDeployment resource will contain owner references to the parent PlatformOperator custom resource the manager is reconciling. The failed deployment of a bundle should be reflected in the aggregate ClusterOperator resource.
+
+Admins may elect to install or delete a PO package after cluster creation. This process would include creating, or deleting, a PlatformOperator custom resource. In order to delete a PO after cluster creation, admins would need to delete the top-level PlatformOperator custom resource for that package.
+
+**Note**: Deleting the top-level PlatformOperator custom resource will result in the cascading deletion of bundle contents. This may result in Kubernetes garbage collecting cluster-scoped resources, e.g. CRDs, which can impact workload health in unsupported configurations. This behavior differs from the current OLM 0.x behavior, which has an opinionated tenancy model it needs to protect, and can lead to data loss scenarios.
+
+In the case that a bundle version has been rolled out to the cluster that's vulnerable, or broken in some capacity, admins may perform the following manual steps to upgrade the PO package version:
+
+- Find the current bundle version for the configured PO package name.
+- Verifying that the redhat-operators index image contains a newer, higher semver bundle than what's currently applied to the cluster.
+- Deleting the underlying bundle deployment resources to force the PO manager to select a new bundle.
+
+These steps may result in workload disruption while new bundle contents are rolled out on the cluster and Kubernetes garbage collects the old bundle contents, and this behavior may require explicit documentation.
```

Comment:
Coming back to this: deleting the underlying workload just to perform a manual upgrade seems like a non-starter given the plain provisioner supports pivoting bundle contents between Bundles without first deleting those resource. 

Maybe having a spec.Unmanaged field on the PlatformOperators resource can help smooth the UX during phase 0. In this model, admins would be responsible for setting the PO resource to an unmanaged state, and then directly editing the underlying BundleDeployment with the newest bundle container image.

Reply:
A `spec.Unmanaged` field would have some complex mechanics when we begin to support dependencies and conflicts. What would it mean for one PO to be managed and another to be unmanaged in the context of a global resolution?

We know we'll need to have a single global resolution to ensure that we've accounted for all of the constraints of all of the requested operators and their dependency trees, so "unmanaged" seems like it could be:
1. the unmanaged PO must still be "mandatory" from the perspective of the resolver so that we ensure its dependencies are present.
2. the thing reading the resolution result and creating/updating the resolveset explicitly disassociates the unmanaged POs existing BD from the resolveset and then leaves the unmanaged PO bundle out of the new resolveset.
3. If the PO is switched back to managed, the thing creating/updating the resolveset re-associates the BD with the resolveset and then ensures that the contents of that BD match what is dictated by the resolution (i.e. anything that was overridden while it was unmanaged gets reverted, and things are put back to the catalog/resolution view)

Another problem with unmanaged in the context of operators and upgrades is that catalog maintainers essentially lose all control of the upgrades that occur while the operator is unmanaged. So a cluster admin could theoretically:
1. Set a PO to unmanaged
2. Make a bunch of changes to the underlying BD that cause various pivots to happen
3. Set the PO back to managed, potentially causing another pivot to get back to "normal"

But steps 2 and 3 are completely unknown/unsupported upgrade paths. At a minimum we'd need to have super clear docs that putting a PO into the unmanaged state essentially voids the warranty on the cluster.

Yet _another_ problem with unmanaged POs is this:
1. Create PO_A, let it reconcile, then set it to unmanaged. You've got PO_A v1 installed now.
2. Create PO_B (PO_B depends on a PO_A == v2, but PO_A is unmanaged and is installed at v1).

What do we do here?

I feel like the "unmanaged" mode has to be "turn off resolution entirely and manage all BDs directly".

Reply:
My feeling is that we just ignore this whole problem in phase 0. If a user gets into a situation where:
- their cluster is vulnerable, they should probably invoke the operator hotfix procedure that the pipeline supports.
- the PO operator is buggy, they should open an issue with the relevant team,
- the PO functionality itself is buggy, they should open an issue with us.

If all else fails in any of those cases, I expect its reasonable for us to say “pull the plug on the cluster” or “delete the PO and all relevant BDs” while we’re in tech preview.

Reply:
See [the tmp-platform-operators](https://coreos.slack.com/archives/C030W4V9BHN/p1658424971405029?thread_ts=1658424971.405029&cid=C030W4V9BHN) thread where we discuss this in more detail.

I can update the EP with this limitation in the next round of edits.